nuxeo-features/nuxeo-automation/nuxeo-automation-core/src/main/java/org/nuxeo/ecm/automation/core/operations/document/AddPermission.java

```diff
@@ -81,88 +95,111 @@
-        }
-
-        String username;
-        if (user == null) {
-            // share a document with someone not registered in Nuxeo, by using only an email
-            username = NuxeoPrincipal.computeTransientUsername(email);
-        } else {
-            username = user;
-            ConfigurationService configService = Framework.getService(ConfigurationService.class);
-            if (configService.isBooleanPropertyFalse(ALLOW_VIRTUAL_USER)) {
-                checkUserExistence(username);
-            }
-        }
-
         ACP acp = doc.getACP() != null ? doc.getACP() : new ACPImpl();
         Map<String, Serializable> contextData = new HashMap<>();
         contextData.put(NOTIFY_KEY, notify);
         contextData.put(COMMENT_KEY, comment);
 
         String creator = session.getPrincipal().getName();
-        ACE ace = ACE.builder(username, permission)
-                     .creator(creator)
-                     .begin(begin)
-                     .end(end)
-                     .contextData(contextData)
-                     .build();
         boolean permissionChanged = false;
         if (blockInheritance) {
             permissionChanged = acp.blockInheritance(aclName, creator);
         }
-        permissionChanged = acp.addACE(aclName, ace) || permissionChanged;
+
+        for (String username : users) {
+            ACE ace = ACE.builder(username, permission)
+                         .creator(creator)
+                         .begin(begin)
+                         .end(end)
+                         .contextData(contextData)
+                         .build();
+            permissionChanged = acp.addACE(aclName, ace) || permissionChanged;
+        }
+
         if (permissionChanged) {
             doc.setACP(acp, true);
         }
     }
 
-    protected void checkUserExistence(String username) {
-        UserManager userManager = Framework.getService(UserManager.class);
-        if (userManager.getUserModel(username) == null && userManager.getGroupModel(username) == null) {
-            String errorMsg = "User or group name '" + username + "' does not exist. Please provide a valid name.";
+    protected void validateParameters() {
+        if (user == null && (users == null || users.isEmpty()) && email == null) {
+            throw new IllegalParameterException("'users' or 'email' parameters must be set");
+        } else if (email != null && end == null) {
+            throw new IllegalParameterException("'end' parameter must be set when adding a permission for an 'email'");
+        }
+
+        ConfigurationService configService = Framework.getService(ConfigurationService.class);
+        if (configService.isBooleanPropertyFalse(ALLOW_VIRTUAL_USER)) {
+            UserManager userManager = Framework.getService(UserManager.class);
+            if (user != null) {
+                checkUserAndGroupExistence(user, userManager);
+            }
+            if (users != null && !users.isEmpty()) {
```

Comment:
`&& !users.isEmpty()` is not needed, there's little cost to doing a loop on an empty list.